PIC_database_updater.py

```diff
@@ -52,49 +59,71 @@ def update_accession(self, barcode, accession):
 
-        sql = self.sql_csv_tools.create_update_statement(tab_name='collectionobject', col_list=['AltCatalogNumber'],
-                                                         val_list=[accession], condition=condition)
+        if pd.isna(is_present) or self.force_update:
+            condition = f"""WHERE CatalogNumber = {barcode};"""
 
+            sql = self.sql_csv_tools.create_update_statement(tab_name='collectionobject', col_list=['AltCatalogNumber'],
+                                                             val_list=[accession], condition=condition)
 
-        self.sql_csv_tools.insert_table_record(sql=sql, logger_int=self.logger)
+            self.sql_csv_tools.insert_table_record(sql=sql)
+        else:
+            self.logger.info(f"Accession number already in collectionobject table at: {barcode}")
 
     def update_herbarium_code(self, barcode, herb_code):
         """function to update herbarium abbreviation code in Modifier column e.g CAS , DS etc...
         args:
             barcode = barcode of record you want to update.
             herb_code = the herbarium acronymn
         """
+        is_present = self.sql_csv_tools.get_one_match(tab_name="collectionobject", id_col="Modifier",
+                                                      key_col="CatalogNumber", match=f"{barcode}")
+        if pd.isna(is_present) or self.force_update:
 
-        condition = f"""WHERE CatalogNumber = "{barcode}";"""
+            condition = f"""WHERE CatalogNumber = {barcode};"""
 
-        sql = self.sql_csv_tools.create_update_statement(tab_name='collectionobject', col_list=['Modifier'],
-                                                         val_list=[herb_code], condition=condition)
+            sql = self.sql_csv_tools.create_update_statement(tab_name='collectionobject', col_list=['Modifier'],
+                                                             val_list=[herb_code], condition=condition)
 
-        self.sql_csv_tools.insert_table_record(sql=sql)
+            self.sql_csv_tools.insert_table_record(sql=sql)
+        else:
+            self.logger.info(f"Herbarium code already in collectionobject table at:{barcode}")
 
 
-    def update_coords(self, barcode, column_list, colname_list):
+    def update_coords(self, barcode, val_list, colname_list):
         """function used to update lat/long columns locality table
             args:
                 barcode: the barcode of the record you want ot update
                 colname_list: the list of database column names to update with locality info.
-                column_list: the list of values to update the locality table with"""
+                column_list: the list of values to update the locality table with
+        """
 
         collecting_event_id = self.get_collectingevent_id(barcode=barcode)
 
         locality_id = self.get_locality_id_with_collectingevent(collecting_event_id=collecting_event_id)
 
```

Comment:
for groups of fields that are almost always updated together like lat/long, the database check will check one of them, on the assumption that both will be empty if one is. In order to avoid doing a ton of costly database checks for every row.